_posts/2021-03-17-encoding-effects-with-freer-simple.markdown

```diff
@@ -0,0 +1,474 @@
+
+- *Built-in* effect in this example is `AppError`
+
+#### Custom Effects
+
+In Custom Effects you need to provide the definition of the **GADT** that is going to describe the Algebra of your Effect and the function that is going to introduce or `send` that Algebra to the `Eff r a` *Monad* that is going to contain the indexed **Open Union Type**, injecting this Algebra into the index list.
+
+> NOTE: It is important to point out that **freer-simple** provides [Template Haskell](https://wiki.haskell.org/Template_Haskell) tooling for auto-generate the `send` injecting
+functions by us, but I've preferred to be a little more boilerplate for educational purpose.
+
+#### DocStorage Effect
+{% highlight haskell %}
+import Control.Monad.Freer as F
+
+data DocStorage r where
+  UploadDocument  :: Document -> DocStorage DocRef
+
+uploadDoc :: Member DocStorage effs => Document -> Eff effs DocRef
+uploadDoc = F.send . UploadDocument
+{% endhighlight %}
+
+#### DataAccess Effect
+{% highlight haskell %}
+import Control.Monad.Freer as F
+
+data DataAccess r where
+  UserExists          :: UserId -> DataAccess Bool
+  SetDocumentUploaded :: Document -> DocRef -> DataAccess ()
+
+userExists :: Member DataAccess effs => UserId -> Eff effs Bool
+userExists = F.send . UserExists
+
+setDocumentUploaded :: Member DataAccess effs => Document -> DocRef -> Eff effs ()
+setDocumentUploaded doc = F.send . SetDocumentUploaded doc
+{% endhighlight %}
+
+#### Logging Effect
+{% highlight haskell %}
+import Control.Monad.Freer as F
+
+data Logging r where
+  LogInfo   :: Text -> Logging ()
+  LogError  :: Text -> Logging ()
+
+logInfo :: Member Logging effs => Text -> Eff effs ()
+logInfo = F.send . LogInfo
+
+logError :: Member Logging effs => Text -> Eff effs ()
+logError = F.send . LogError
+{% endhighlight %}
+
+#### Built-In Effects
+In the case of *built-in* effects the effort is minimum because the Effect and handler is already provided by the library and we only need to define some aliases, but only if we want as in my case.
+
+{% highlight haskell %}
+{-# LANGUAGE PatternSynonyms #-}
+
+import Control.Monad.Freer.Error
+
+data Err 
+  = NotFound Text
+  | InternalError Text
+  | Forbidden Text
+  | BadRequest Text
+  deriving (Eq, Show)
+
+pattern ERROR_UPLOADING_DOCUMENT
+      , DOCUMENT_NOT_FOUND
+      , USER_NOT_FOUND
+     :: Err
+pattern ERROR_UPLOADING_DOCUMENT   = InternalError "Document uploading error"
+pattern DOCUMENT_NOT_FOUND         = NotFound      "Document was not found on the Storage"
+pattern USER_NOT_FOUND             = NotFound      "User was not found on the system"
+
+type AppError = Error Err
+{% endhighlight %}
+
+In this case the library already provide a **Higher Order Type** which is `newtype Error e r` which Value Constructor is `Error :: e -> Error e r`, therefore its kind is `* -> * -> *`. Our alias is `type AppError = Error Err` whose kind is `* -> *` which coincide with the Kind that is contain by the Indexed **Open Union Type**
```

Comment:
Recall that Star "*" is deprecated as a kind since GHC 8.X.X (I don't recall)

Reply:
No, it is not. I am using 8.6.5 and it is not deprecated. You need to disable `StarIsType` Language extension for having `Type -> Type -> Type`